File: e2e/native/java/src/test/java/e2e/AnvilTest.java
package e2e;

import com.example.contracts.Token;
import java.math.BigInteger;
import org.junit.jupiter.api.Test;
import org.web3j.crypto.Credentials;
import org.web3j.protocol.Web3j;
import org.web3j.protocol.core.DefaultBlockParameterName;
import org.web3j.protocol.core.methods.response.TransactionReceipt;
import org.web3j.protocol.http.HttpService;
import org.web3j.tx.RawTransactionManager;
import static org.junit.jupiter.api.Assertions.*;

class AnvilTest {
    @Test void signedTransactionsReadBackAndDecodeEvents() throws Exception {
        String url = System.getenv("ATG_RPC_URL");
        if (url == null) return;
        String contract = System.getenv("ATG_TOKEN_ADDRESS");
        var credentials = Credentials.create(System.getenv("ATG_PRIVATE_KEY"));
        long chainId = Long.parseLong(System.getenv("ATG_CHAIN_ID"));
        var web3j = Web3j.build(new HttpService(url));
        try {
            var client = new Token.Client(contract, web3j, new RawTransactionManager(web3j, credentials, chainId));
            var options = new Token.TransactionOptions(web3j.ethGasPrice().send().getGasPrice(), BigInteger.valueOf(300_000));
            String owner = credentials.getAddress();
            var amount = BigInteger.valueOf(41);
            var mint = Token.sendMint(client, owner, amount, options);
            var mintReceipt = receipt(web3j, mint.getTransactionHash());
            assertEquals(BigInteger.ONE, new BigInteger(mintReceipt.getStatus().substring(2), 16));
            assertEquals(amount, Token.callBalanceOf(client, owner));
            var transfer = mintReceipt.getLogs().stream().map(Token::decodeTransferEvent).filter(java.util.Objects::nonNull).findFirst().orElseThrow();
            assertEquals(amount, transfer.amount());
            var filter = Token.filterTransferEvent(contract, DefaultBlockParameterName.EARLIEST, DefaultBlockParameterName.LATEST, null, owner);
            assertFalse(web3j.ethGetLogs(filter).send().getLogs().isEmpty());

            var approve = Token.sendApprove(client, owner, BigInteger.TEN, options);
            var approveReceipt = receipt(web3j, approve.getTransactionHash());
            assertEquals(BigInteger.TEN, Token.callAllowance(client, owner, owner));
            var approval = approveReceipt.getLogs().stream().map(Token::decodeApprovalEvent).filter(java.util.Objects::nonNull).findFirst().orElseThrow();
            assertEquals(BigInteger.TEN, approval.amount());
        } finally {
            web3j.shutdown();
        }
    }
    private TransactionReceipt receipt(Web3j web3j, String hash) throws Exception {
        for (int attempt = 0; attempt < 30; attempt++) {
            var found = web3j.ethGetTransactionReceipt(hash).send().getTransactionReceipt();
            if (found.isPresent()) return found.get();
            Thread.sleep(100);
        }
        throw new IllegalStateException("transaction receipt not available: " + hash);
    }
}

File: e2e/native/kotlin/src/test/kotlin/AnvilTest.kt
import com.example.contracts.Token
import java.math.BigInteger
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import org.web3j.crypto.Credentials
import org.web3j.protocol.Web3j
import org.web3j.protocol.core.DefaultBlockParameterName
import org.web3j.protocol.core.methods.response.TransactionReceipt
import org.web3j.protocol.http.HttpService
import org.web3j.tx.RawTransactionManager

class AnvilTest {
    @Test
    fun signedTransactionsReadBackAndDecodeEvents() {
        val url = System.getenv("ATG_RPC_URL") ?: return
        val contract = requireNotNull(System.getenv("ATG_TOKEN_ADDRESS"))
        val credentials = Credentials.create(requireNotNull(System.getenv("ATG_PRIVATE_KEY")))
        val chainId = requireNotNull(System.getenv("ATG_CHAIN_ID")).toLong()
        val web3j = Web3j.build(HttpService(url))
        try {
            val manager = RawTransactionManager(web3j, credentials, chainId)
            val client = Token.Client(contract, web3j, manager)
            val gasPrice = web3j.ethGasPrice().send().gasPrice
            val options = Token.TransactionOptions(gasPrice, BigInteger.valueOf(300_000))
            val owner = credentials.address
            val amount = BigInteger.valueOf(41)
            val mint = Token.sendMint(client, owner, amount, options)
            val mintReceipt = receipt(web3j, requireNotNull(mint.transactionHash))
            assertEquals(BigInteger.ONE, mintReceipt.status?.removePrefix("0x")?.toBigInteger(16))
            assertEquals(amount, Token.callBalanceOf(client, owner))
            val transfer = mintReceipt.logs.firstNotNullOfOrNull(Token::decodeTransferEvent)
            assertNotNull(transfer)
            assertEquals(amount, transfer.amount)
            val filtered = Token.filterTransferEvent(contract, DefaultBlockParameterName.EARLIEST, DefaultBlockParameterName.LATEST, null, owner)
            assertTrue(web3j.ethGetLogs(filtered).send().logs.isNotEmpty())

            val approve = Token.sendApprove(client, owner, BigInteger.TEN, options)
            val approveReceipt = receipt(web3j, requireNotNull(approve.transactionHash))
            assertEquals(BigInteger.TEN, Token.callAllowance(client, owner, owner))
            val approval = approveReceipt.logs.firstNotNullOfOrNull(Token::decodeApprovalEvent)
            assertNotNull(approval)
            assertEquals(BigInteger.TEN, approval.amount)
        } finally {
            web3j.shutdown()
        }
    }

    private fun receipt(web3j: Web3j, hash: String): TransactionReceipt {
        repeat(30) {
            val response = web3j.ethGetTransactionReceipt(hash).send()
            if (response.transactionReceipt.isPresent) return response.transactionReceipt.get()
            Thread.sleep(100)
        }
        error("transaction receipt not available: $hash")
    }
}
